### PR TITLE
feat: CSRF plugin

### PR DESCRIPTION
**Pull request by o0Ignition0o**
This Pull request does a couple of things:

- Add a CSRF protection plugin, that is now enabled by default.
- Changes the default CORS `allow_headers` behavior from allow only `Content-Type`, `apollographql-client-name` and `apollographql-client-version` to mirror the received `access-control-request-headers`

Co-authored-by: Simon Sapin <simon@apollographql.com>

### CODE REVIEW

**Review by glasser — CHANGES_REQUESTED** [review]
I didn't expect you to implement this within hours of my incoherent "omg I just found an issue but also have to pick up my daughter at school in 7 minutes" ramblings — wow, thanks!

File: apollo-router-core/src/plugins/csrf.rs [hunk 1]
```diff
@@ -0,0 +1,289 @@
...
+
+fn should_accept(req: &RouterRequest) -> bool {
+    let headers = req.originating_request.headers();
+    headers_require_preflight(headers) || content_type_requires_preflight(headers)
+}
+
+fn headers_require_preflight(headers: &HeaderMap) -> bool {
+    headers
+        .keys()
+        .any(|header_name| !NON_PREFLIGHTED_HEADER_NAMES.contains(header_name))
```
[the remark this answers is not in the document]
Comment (o0Ignition0o) [reply apollo-router-core/src/plugins/csrf.rs:88]: `he spec says that we need to preflight if the JS code tries to add another header` oh i did not catch this part, thanks a lot!

File: apollo-router-core/src/plugins/csrf.rs [hunk 2]
```diff
@@ -0,0 +1,289 @@
...
+}
+
+fn content_type_requires_preflight(headers: &HeaderMap) -> bool {
+    headers
+        .get(header::CONTENT_TYPE)
```
[the remark this answers is not in the document]
Comment (o0Ignition0o) [reply apollo-router-core/src/plugins/csrf.rs:93]: https://fetch.spec.whatwg.org/#cors-safelisted-request-header claims 
`This intentionally does not use [extract a MIME type](https://fetch.spec.whatwg.org/#concept-header-extract-mime-type) as that algorithm is rather forgiving and servers are not expected to implement it.`

The example seems to concur, so I'm tempted to:

- use `get_all`
- bail if there are more than one value
- not parse the mime type and check against string representation (as it looks like what the spec does)
wdyt?

File: apollo-router-core/src/plugins/csrf.rs [hunk 3]
```diff
@@ -0,0 +1,289 @@
...
+        .map(|content_type| {
+            if let Ok(as_str) = content_type.to_str() {
+                if let Ok(mime_type) = as_str.parse::<mime::Mime>() {
```
[the remark this answers is not in the document]
Comment (o0Ignition0o) [reply apollo-router-core/src/plugins/csrf.rs:96]: I'm going to keep a close eye on the new crate, which seems to be going in the right direction.

I agree, let's:
- treat failure to turn the HeaderValue into a valid utf8 string as a condition that would have triggered preflight
- trim and replace \t with ' ' before we try to parse
- treat failure to parse mime type as a non sufficient condition to trigger a preflight

[the remark this answers is not in the document]
Comment (o0Ignition0o) [reply apollo-router-core/src/plugins/csrf.rs]: should be fixed in [331a5e1](https://github.com/apollographql/router/pull/1006/commits/331a5e18e6cc20858b0d0ffd28719e930d178d6b) but i ll add some tests to make sure it s the case

File: apollo-router-core/src/plugins/csrf.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (o0Ignition0o) [reply apollo-router-core/src/plugins/csrf.rs]: should be fixed in [331a5e1](https://github.com/apollographql/router/pull/1006/commits/331a5e18e6cc20858b0d0ffd28719e930d178d6b) but i ll add some tests to make sure it s the case

**Review by garypen — COMMENTED** [review]
I don't know enough about how CSRF prevention should be implemented to approve, but I've asked a couple of questions...

File: apollo-router-core/src/plugins/csrf.rs [hunk 4]
```diff
@@ -0,0 +1,320 @@
...
+
+fn recommended_header_is_provided(headers: &HeaderMap, recommended_headers: &[String]) -> bool {
+    recommended_headers
+        .iter()
+        .any(|header| headers.get(header).is_some())
+}
+
+fn content_type_requires_preflight(headers: &HeaderMap) -> bool {
+    headers
+        .get_all(header::CONTENT_TYPE)
```
[the remark this answers is not in the document]
Comment (o0Ignition0o) [reply apollo-router-core/src/plugins/csrf.rs:106]: That s a good question, I'll add more comments.

The specification claims a `non simple request` made by a browser needs preflight.
A request is considered [simple](https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS#simple_requests) if all three points below apply:
- the method is either `GET` `POST` or `HEAD`
- the only headers added by javascript code are part of the cors safelisted request headers ([Accept](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Accept),Accept-Language,[Content-Language](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Language),[Content-Type](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Type), and simple [Range](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Range)
- if content-type is set, it must be with a mime type that is `application/x-www-form-urlencoded` OR `multipart/form-data` OR `text/plain`

A CSRF mitigation mechanism basically makes sure any request made by a browser went through the CORS dance (also known as preflight).
So what we're trying to make sure of in the CSRF plugin is that the request that was sent is not a simple one.

There are two ways we can make sure of that:
- content-type is not one of the three safe essences mentioned above
- a custom header has been sent in the query, and we know for sure it wasnt done by the browser.

```rust
fn should_accept(req: &RouterRequest, recommended_headers: &[String]) -> bool {
    let headers = req.originating_request.headers();
    content_type_requires_preflight(headers)
        || recommended_header_is_provided(headers, recommended_headers)
}
```
I should probably add some variants of this comment to the code documentation, and possibly rename `should_accept` to `request_required_preflight`

> I gather that should_accept is looking to continue the flow of execution, but the processing in content_type_requires_preflight returns true (i.e.: continue) at line 125 which means we are going to continue if we can't process the content-type as a string. Is that correct?

This is correct, as explained in the [fetch standard](https://fetch.spec.whatwg.org/#cors-safelisted-request-header):

Headers, such as `Content-Type`, are [byte sequences](https://infra.spec.whatwg.org/#byte-sequence),

and

To get a byte sequence out of a string, using UTF-8 encode from the Encoding Standard is encouraged.

failure to turn a `HeaderValue` into a string means the byte sequence isn't valid UTF-8, so the header can definitely not be a safe one.

Note that the behavior is close to the [typescript implementation](https://github.com/apollographql/apollo-server/blob/12bf5fc8ef305caa6a8848e37f862d32dae5957f/packages/server/src/preventCsrf.ts#L60) except the mime parsing algorithm doesnt quite follow the fetch specification. So failure to turn `HeaderValue` into a string means the header isn't safelisted (so the request has been preflighted), but failure to parse the header value into a mime type could mean the rust implementation differs from the fetch specification, so we can't safely assume the request was preflighted.

File: apollo-router-core/src/plugins/csrf.rs [hunk 5]
```diff
@@ -0,0 +1,320 @@
...
+use tower::{BoxError, ServiceBuilder, ServiceExt};
+
+#[derive(Deserialize, Debug, Clone, JsonSchema)]
+#[serde(deny_unknown_fields)]
+struct CSRFConfig {
+    /// CSRFConfig is enabled by default;
+    /// set disabled = true to disable the plugin behavior
+    #[serde(default)]
+    disabled: bool,
+    /// Override the headers to check for by setting
```
[the remark this answers is not in the document]
Comment (glasser) [reply apollo-router-core/src/plugins/csrf.rs:17]: CSRF prevention = block clients (most notably GET requests) that don't jump through certain hoops. At the very least, this might require a bit of work for users when they upgrade to this version for the first time.

File: apollo-router-core/src/plugins/csrf.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (glasser) [reply apollo-router-core/src/plugins/csrf.rs]: @o0Ignition0o I'll look at the code in a sec, but I don't think "parse and get the essence of each value" is what I'm asking for. The fetch spec wants you to parse just one content-type — it's just that that content-type should be formed by combining all content-type headers rather than by just arbitrarily picking one of them (which is what `get()` does)

[the remark this answers is not in the document]
Comment (o0Ignition0o) [reply apollo-router-core/src/plugins/csrf.rs]: can you please point me to the spec? I can't seem to find the "combine all values" part (and all i can find is how to parse one mime type https://mimesniff.spec.whatwg.org/#parse-a-mime-type)

**Review by glasser — CHANGES_REQUESTED** [review]

File: apollo-router-core/src/plugins/csrf.rs [hunk 6]
```diff
@@ -0,0 +1,320 @@
...
+    /// custom_headers
+    /// Note that if you set recommended_headers here,
+    /// you may also want to have a look at your `CORS` configuration,
+    /// and make sure you either:
+    /// - did not set any `allow_headers` list (so it defaults to `mirror_request`)
+    /// - added your recommended headers to the allow_headers list, as shown in the
+    /// `examples/cors-and-csrf/*.router.yaml` files.
+    #[serde(default = "apollo_custom_preflight_headers")]
+    recommended_headers: Vec<String>,
```
[the remark this answers is not in the document]
Comment (o0Ignition0o) [reply apollo-router-core/src/plugins/csrf.rs:26]: `required_headers` it is! 🚀

File: apollo-router-core/src/plugins/csrf.rs [hunk 7]
```diff
@@ -0,0 +1,320 @@
...
+                    // a valid mime type... which would be safe enough for us to assume preflight was triggered if the `mime`
+                    // crate followed the fetch specification, but it unfortunately doesn't (see comment above).
+                    //
+                    // Better safe than sorry, we will mark this content_type value as not sufficient to have triggered preflight
+                    false
+                }
+            } else {
+                // If we get here, this means that turning the content-type header value
+                // into a string failed (ie it's not valid UTF-8) which would lead to a preflight.
+                true
```
[the remark this answers is not in the document]
Comment (o0Ignition0o) [reply apollo-router-core/src/plugins/csrf.rs:125]:
```
        // We couldn't parse a header value, let's err on the side of caution here
```
looks like a more sensible comment as well so let s go with that!

File: apollo-router-core/src/plugins/csrf.rs [hunk 8]
```diff
@@ -0,0 +1,329 @@
...
+        .collect::<Result<Vec<_>, _>>()
+    {
+        // https://fetch.spec.whatwg.org/#concept-header-list-combine
+        combined_headers.join("\u{002C}\u{0020}") // ', '
+    } else {
+        // We couldn't parse a header value, let's err on the side of caution here
+        return false;
+    };
+
+    dbg!("'\u{002C}\u{0020}'");
```
Comment (glasser) [inline apollo-router-core/src/plugins/csrf.rs:123]: probably remove this?

**Review by glasser — APPROVED** [review]

File: apollo-router-core/src/plugins/csrf.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (garypen) [reply apollo-router-core/src/plugins/csrf.rs]: Great explanation @o0Ignition0o ! If some of that makes it into the code, then it will serve us well in the future.

[the remark this answers is not in the document]
Comment (o0Ignition0o) [reply apollo-router-core/src/plugins/csrf.rs]: glad you commented! The "what is csrf" part is now documented [here](https://github.com/apollographql/router/blob/00769c2af6fee1d5c7010ac736744d58d8644b14/apollo-router-core/src/plugins/csrf.rs#L42) and the details about the algorithm are commented respectively in [is_preflighted](https://github.com/apollographql/router/blob/00769c2af6fee1d5c7010ac736744d58d8644b14/apollo-router-core/src/plugins/csrf.rs#L116), [content_type_requires_preflight](https://github.com/apollographql/router/blob/00769c2af6fee1d5c7010ac736744d58d8644b14/apollo-router-core/src/plugins/csrf.rs#L135) and [recommended headers is provided](https://github.com/apollographql/router/blob/00769c2af6fee1d5c7010ac736744d58d8644b14/apollo-router-core/src/plugins/csrf.rs#L172)

File: CONTRIBUTING.md [hunk 9]
```diff
@@ -61,14 +61,17 @@ and [Apollo's Docs Theme for Gatsby](https://github.com/apollographql/gatsby-the
...
+and update this file
 
 To build and run the documentation site locally, you'll have to install the relevant packages by doing the following
 from the root of the `router` repository:
 
 ```sh
 cd docs
 npm i
-npm start
+npm start -->
```
Comment (abernix) [inline CONTRIBUTING.md:74]: 🤔

File: docs/source/configuration/cors.mdx [hunk 10]
```diff
@@ -1,10 +1,210 @@
 ---
```
Comment (abernix) [inline docs/source/configuration/cors.mdx:1]: Let's separate out documentation to another PR?

File: apollo-router-core/src/plugins/csrf.rs [hunk 11]
```diff
@@ -0,0 +1,335 @@
...
+                        let error = crate::Error {
+                            message: format!("This operation has been blocked as a potential Cross-Site Request Forgery (CSRF). \
+                            Please either specify a 'content-type' header (with a mime-type that is not one of {}) \
+                            or provide one of the following headers: {}", 
+                            NON_PREFLIGHTED_CONTENT_TYPES.join(", "),
+                            required_headers.join(", ")),
+                            locations: Default::default(),
+                            path: Default::default(),
+                            extensions: Default::default(),
+                        };
```
Comment (bnjjj) [inline apollo-router-core/src/plugins/csrf.rs:109]: NIT: If think we have builders for errors

File: CONTRIBUTING.md (hunk 9 above)
Comment (o0Ignition0o) [reply CONTRIBUTING.md:74]: the three steps defined above:

```
cd docs
npm i
npm start``` are out of date; We will turn this //TODO comment into an issue, which @SimonSapin wants to tackle as a followup

**Review by bnjjj — APPROVED** [review]

File: CONTRIBUTING.md (hunk 9 above)
Comment (SimonSapin) [reply CONTRIBUTING.md:74]: I’ve submitted https://github.com/apollographql/router/pull/1036. Maybe undo this chunk here to avoid a merge conflict?

File: apollo-router-core/src/plugins/csrf.rs (hunk 11 above)
Comment (o0Ignition0o) [reply apollo-router-core/src/plugins/csrf.rs]: nice catch! fixed in a20105b4

**Review by Geal — APPROVED** [review]
